Add Diagonal and Mask Operators to allow inpainting

jakobsj opened the pull request:
Adds two new operators, DiagonalOperator and MaskOperator. DiagonalOperator is a subclass of LinearOperator and represents a diagonal operator by taking as input a DataContainer/ImageData/Acquisition data, which will simply by multiplied onto an input to direct or adjoint. MaskOperator is a special case with the diagonal being 1/0-valued.

Unit tests added, and builds locally.

Closes #321, closes #462.

paskino commented on Wrappers/Python/ccpi/optimisation/operators/DiagonalOperator.py:
```diff
@@ -0,0 +1,94 @@
+        if out is None:
+            return self.diagonal * x
+        else:
+            out.fill(self.diagonal * x)
```
```python
self.diagonal.muliply(x, out=out)
```

paskino commented on Wrappers/Python/ccpi/optimisation/operators/DiagonalOperator.py:
```diff
@@ -0,0 +1,94 @@
+# -*- coding: utf-8 -*-
+#  CCP in Tomographic Imaging (CCPi) Core Imaging Library (CIL).
+
+#   Copyright 2017 UKRI-STFC
```
I think it's 2020 ;)

jakobsj replied:
Yeah I wasn't sure whether to leave it as 2017 or change to current year. Somehow I don't like all the different files to have different years on from which year they happened to be written. I'd rather have an identical header in all files with the same year shown referring to the year of the software release perhaps.

epapoutsellis replied:
Agree changing all to 2020, or maybe 2017-2020. It will be ok with our pubs.

paskino commented on Wrappers/Python/ccpi/optimisation/operators/DiagonalOperator.py:
```diff
@@ -0,0 +1,94 @@
+        if out is None:
+            return self.diagonal * x
+        else:
+            out.fill(self.diagonal * x)
```
as `direct`

epapoutsellis replied:
if you want to reduce lines 
` return self.direct(x, out=out)`

paskino replied:
This is actually better for code maintenance and clarity.

paskino commented on Wrappers/Python/ccpi/optimisation/operators/DiagonalOperator.py:
```diff
@@ -0,0 +1,94 @@
+        
+        '''Evaluates operator norm of DiagonalOperator'''
+        
+        return self.diagonal.as_array().max()
```
`max` is a method of `DataContainer` so you could use it directly

```python 
return self.diagonal.max()
```

paskino commented on Wrappers/Python/test/test_Operator.py:
```diff
@@ -110,7 +111,48 @@ def test_ScaledOperator(self):
+        
+        M = 3
+        ig = ImageGeometry(M, M)
+        x = ig.allocate('random_int')
```
please seed the random number generator with `numpy.random.seed(1)` or use the `seed=1` argument of `allocate`.

Also, since in #535 the type returned by `random_int` has changed, I'd suggest to use `random`.

paskino commented on Wrappers/Python/test/test_Operator.py:
```diff
@@ -110,7 +111,48 @@ def test_ScaledOperator(self):
+        M = 3
+        ig = ImageGeometry(M, M)
+        x = ig.allocate('random_int')
+        diag = ig.allocate('random_int')
```
idem

paskino commented on Wrappers/Python/test/test_Operator.py:
```diff
@@ -110,7 +111,48 @@ def test_ScaledOperator(self):
+        
+        M = 3
+        ig = ImageGeometry(M, M)
+        x = ig.allocate('random_int')
```
In this case it makes sense that the mask is of `int` type, see #535.

jakobsj replied:
I actually wanted the mask to be of Boolean type rather than numerical 1.0 and 0.0, but didn't know whether that was possible.

paskino replied:
yes, you can use `dtype=numpy.bool` when allocating.

paskino commented:
At any rate, Jenkins was very happy.

epapoutsellis commented on Wrappers/Python/ccpi/optimisation/operators/MaskOperator.py:
```diff
@@ -0,0 +1,157 @@
+        super(MaskOperator, self).__init__(mask)
+        
+
+if __name__ == '__main__':
```
After merging to master, added in cil-demos and maybe in the tests directory

epapoutsellis commented on Wrappers/Python/ccpi/optimisation/operators/DiagonalOperator.py:
```diff
@@ -0,0 +1,94 @@
+    operator. Maps an element of :math:`x\in X` onto the element 
+    :math:`y \in X,  y = diag*x`, where * denotes elementwise multiplication.
+                       
+        :param diagonal: DataContainer with diagonal elements
```
I think in this context `diagonal elements` is a bit confusing, unless you add sth like
![image](https://user-images.githubusercontent.com/22398586/78279201-18d82880-750f-11ea-95d8-e24ce47fff78.png)
in the description.

jakobsj replied:
Or are you happy with the operator being called DiagonalOperator, but wanting the "diagonal elements" to be referred to differently?

jakobsj replied:
Added more explanation. Not sure about the continuation code, maybe can add once LinearOperatorMatrix is ready?

jakobsj commented:
Yes I can see that. Maybe DiagonalOperator is not the best name? Not sure
what else to call it though. ElementwiseOperator? Weight(ing)Operator?

On Thu, 2 Apr 2020, 19:24 Vaggelis Papoutsellis, <notifications@github.com>
wrote:

> *@epapoutsellis* commented on this pull request.
> ------------------------------
>
> In Wrappers/Python/ccpi/optimisation/operators/DiagonalOperator.py
> <https://github.com/vais-ral/CCPi-Framework/pull/538#discussion_r402482794>
> :
>
> > +from __future__ import absolute_import
> +from __future__ import division
> +from __future__ import print_function
> +
> +import numpy as np
> +from ccpi.framework import ImageData
> +from ccpi.optimisation.operators import LinearOperator
> +
> +class DiagonalOperator(LinearOperator):
> +
> +    r'''DiagonalOperator:  D: X -> X,  takes in a DataContainer or subclass
> +    thereof, diag, representing elements on the diagonal of a diagonal
> +    operator. Maps an element of :math:`x\in X` onto the element
> +    :math:`y \in X,  y = diag*x`, where * denotes elementwise multiplication.
> +
> +        :param diagonal: DataContainer with diagonal elements
>
> I think in this context diagonal elements is a bit confusing, unless you
> add sth like
> [image: image]
> <https://user-images.githubusercontent.com/22398586/78279201-18d82880-750f-11ea-95d8-e24ce47fff78.png>
> in the description.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vais-ral/CCPi-Framework/pull/538#pullrequestreview-386638880>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACMDSCAWFWP7JST45AYG2G3RKTC47ANCNFSM4L2LYAGA>
> .
>

jakobsj commented:
OK, I'll add a bit more description.

I'm unsure what to do with the continuation code. It's neat that we can do
the same with the LinearOperatorMatrix, but if that is not yet available,
how about adding that once it is ready, so we can close the current PR now
(when I've made the minor changes suggested?)

On Thu, 2 Apr 2020 at 23:38, Vaggelis Papoutsellis <notifications@github.com>
wrote:

> *@epapoutsellis* commented on this pull request.
> ------------------------------
>
> In Wrappers/Python/ccpi/optimisation/operators/DiagonalOperator.py
> <https://github.com/vais-ral/CCPi-Framework/pull/538#discussion_r402610871>
> :
>
> > +from __future__ import absolute_import
> +from __future__ import division
> +from __future__ import print_function
> +
> +import numpy as np
> +from ccpi.framework import ImageData
> +from ccpi.optimisation.operators import LinearOperator
> +
> +class DiagonalOperator(LinearOperator):
> +
> +    r'''DiagonalOperator:  D: X -> X,  takes in a DataContainer or subclass
> +    thereof, diag, representing elements on the diagonal of a diagonal
> +    operator. Maps an element of :math:`x\in X` onto the element
> +    :math:`y \in X,  y = diag*x`, where * denotes elementwise multiplication.
> +
> +        :param diagonal: DataContainer with diagonal elements
>
> No DiagonalOperator is the right name, either we have it as a map:
> D(x_{1}, ..., x_{n}) = (a_{1}x_{1}, ..., a_{n}x_{n})
> or as a matrix-vector multiplication. Just add few more words in the
> description, maybe mention how to do it as a matrix-vector multiplication.
> The code below is the continuation from your script:
>
> from ccpi.optimisation.operators import LinearOperatorMatrixfrom ccpi.framework import VectorData, DataContainer
>
>
>     diag_mat = np.diag(diag.as_array().flatten('F'))
>     x_vec = x.as_array().flatten('F')
>
>     res = np.reshape(diag_mat.dot(x_vec), x.shape, 'F')
>
>     print("\n Result matrix-vector \n {} \n".format(res))
>     print("Result non  mat_vec \n {}".format(z.as_array()))
>
>
>     L = LinearOperatorMatrix(diag_mat)
>     x_vec_data = VectorData(x_vec)
>
>     res1 = np.reshape(L.direct(x_vec_data).as_array(), x.shape, 'F')
>     print("\nResult with LinearOperatorMatrix \n{}".format(res1))
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vais-ral/CCPi-Framework/pull/538#discussion_r402610871>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACMDSCHX5UTLQ53VFAP3SMTRKUAW3ANCNFSM4L2LYAGA>
> .
>

jakobsj commented:
OK, I believe I have addressed all the comments, please have another look.

Also, in MaskOperator I added in a trivial/alias attribute `mask` simply pointing to the attribute `diagonal`, since the user who creates a MaskOperator with input `mask` and then wants to query what the operator holds, would until now, somewhat surprisingly, have to ask to see the `diagonal`. But can now ask for mask instead, which should be more intuitive.

paskino commented:
I will merge this after I merge #541 which solves the test failure that jenkins found after you merged to master.
